folio-token/src/main/java/org/folio/spring/service/SystemUserService.java

```diff
@@ -0,0 +1,141 @@
+package org.folio.spring.service;
+
+import static java.util.Objects.isNull;
+import static org.folio.spring.utils.TokenUtils.parseUserTokenFromCookies;
+import static org.springframework.http.HttpHeaders.SET_COOKIE;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import java.time.Instant;
+import java.util.Optional;
+import java.util.function.Supplier;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.folio.spring.client.AuthnClient;
+import org.folio.spring.client.AuthnClient.UserCredentials;
+import org.folio.spring.client.UsersClient;
+import org.folio.spring.config.properties.FolioEnvironment;
+import org.folio.spring.context.ExecutionContextBuilder;
+import org.folio.spring.model.SystemUser;
+import org.folio.spring.model.UserToken;
+import org.folio.spring.scope.FolioExecutionContextSetter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+@Log4j2
+@Service
+@RequiredArgsConstructor
+public class SystemUserService {
+
+  private final ExecutionContextBuilder contextBuilder;
+  private final SystemUserProperties systemUserProperties;
+  private final FolioEnvironment environment;
+  private final AuthnClient authnClient;
+  private final PrepareSystemUserService prepareUserService;
+
+  private Cache<String, SystemUser> systemUserCache;
+
+  /**
+   * Get authenticate system user.
+   * Get from cache if present (or loginWithExpiry otherwise). Call loginWithExpiry token endpoint in case
+   * access token expired. Call login endpoint in case refresh token expired
+   *
+   * @param tenantId The tenant name
+   * @return {@link org.folio.spring.model.SystemUser} with token value
+   */
+  public SystemUser getAuthedSystemUser(String tenantId) {
+    if (systemUserCache == null) {
+      return getSystemUser(tenantId);
+    }
+
+    var user = systemUserCache.get(tenantId, this::getSystemUser);
+    var userToken = user.token();
+    var now = Instant.now();
+    if (userToken.accessTokenExpiration().isAfter(now)) {
+      return user;
+    }
+
+    systemUserCache.invalidate(tenantId);
+    if (userToken.accessTokenExpiration().isAfter(now)) {
+      var newToken = login(user.username());
+      user = user.withToken(newToken);
+    } else {
+      user = getSystemUser(tenantId);
+    }
+    systemUserCache.put(tenantId, user);
+
+    return user;
+  }
+
+  /**
+   * Authenticate system user and return token value.
+   *
+   * @param user {@link org.folio.spring.model.SystemUser} to log with
+   * @return token value
+   */
+  public UserToken authSystemUser(SystemUser user) {
+    return getToken(() ->
+            authnClient.loginWithExpiry(new UserCredentials(user.username(), systemUserProperties.password())),
```

Comment:
Like my comment above. This only calls one login endpoint. In this case you're calling `authn/login-with-expiry`. You need to do that, but also if it fails call `authn/login`. Both of these public methods need to do this same check in this sequence.

Reply:
This is done in getSystemUser method

![image](https://github.com/folio-org/folio-spring-support/assets/104345088/377644e7-a715-457c-9e49-6b0d3b3e1503)

Reply:
changes done.